test/functional/rpc_invalidateblock.py

```diff
@@ -35,9 +39,17 @@ def run_test(self):
         self.connect_nodes(0, 1)
         self.sync_blocks(self.nodes[0:2])
         assert_equal(self.nodes[0].getblockcount(), 6)
-        badhash = self.nodes[1].getblockhash(2)
+
+        # Add a header to the tip of node 0 without submitting the block. This shouldn't
+        # affect results since this chain will be invalidated next.
+        tip = int(self.nodes[0].getbestblockhash(), 16)
+        block_time = self.nodes[0].getblock(self.nodes[0].getbestblockhash())['time'] + 1
+        block = create_block(tip, create_coinbase(self.nodes[0].getblockcount()), block_time, version=4)
+        block.solve()
+        self.nodes[0].submitheader(block.serialize().hex())
 
         self.log.info("Invalidate block 2 on node 0 and verify we reorg to node 0's original chain")
+        badhash = self.nodes[1].getblockhash(2)
```

Comment:
It might be a good idea to expand this a bit:
```diff
@@ -45 +45 @@ class InvalidateTest(BitcoinTestFramework):
-        tip = int(self.nodes[0].getbestblockhash(), 16)
+        tip = self.nodes[0].getbestblockhash()
@@ -47 +47 @@ class InvalidateTest(BitcoinTestFramework):
-        block = create_block(tip, create_coinbase(self.nodes[0].getblockcount()), block_time, version=4)
+        block = create_block(int(tip, 16), create_coinbase(self.nodes[0].getblockcount()), block_time, version=4)
@@ -49,0 +50 @@ class InvalidateTest(BitcoinTestFramework):
+        assert_equal(self.nodes[0].getblockchaininfo()["headers"], self.nodes[0].getblockchaininfo()["blocks"] + 1)
@@ -56,0 +58,11 @@ class InvalidateTest(BitcoinTestFramework):
+        assert_equal(self.nodes[0].getblockchaininfo()["headers"], self.nodes[0].getblockchaininfo()["blocks"])
+
+        self.log.info("Reconsider block 6 on node 0 again and verify that the best header is set correctly")
+        self.nodes[0].reconsiderblock(tip)
+        assert_equal(self.nodes[0].getblockchaininfo()["headers"], self.nodes[0].getblockchaininfo()["blocks"] + 1)
+
+        self.log.info("Invalid block 2 on node 0 and verify we reorg to node 0's original chain again")
+        self.nodes[0].invalidateblock(badhash)
+        assert_equal(self.nodes[0].getblockcount(), 4)
+        assert_equal(self.nodes[0].getbestblockhash(), besthash_n0)
+        # Should report consistent blockchain info
```
This additionally checks that the new header can be reconsidered again correctly (`BLOCK_FAILED_CHILD` is indeed reset). It's a bit verbose and repetitive though and following along with the various chainstates is already a bit annoying, so not sure if it delivers enough value.